### PR TITLE
[Quorum Store] txn count-based backpressure

### PR DESCRIPTION
**Pull request by bchocho**
### Description

Add a txn count-based backpressure to quorum store.
* Backpressure is determined by the number of total outstanding txns as seen in the proof_manager. (Note this includes txns in flight in consensus.)
* On backpressure, batch generator reduces by half the max batch size. When no backpressure, batch generator additively adds to the max batch size. (AIMD, similar to TCP.)
* Regardless of backpressure, a batch is created at a regular cadence (after delay of 250ms).

Also:
* Break out the consensus configs to those that are used when quorum store is turned on.
* Add and fix latency metrics (#6775)

### Test Plan

Observe that global TPS and high-gas latency are reasonable. May continue tuning the configs.

### CODE REVIEW

**Comment by bchocho** [comment]
Not quite ready

**Comment by bchocho** [comment]
Tests with quorum store hard-coded to enabled:
* High-gas overload: https://github.com/aptos-labs/aptos-core/pull/6710#issuecomment-1444572809
* Land blocking: https://github.com/aptos-labs/aptos-core/pull/6710#issuecomment-1444619593

File: config/src/config/consensus_config.rs [hunk 1]
```diff
@@ -6,13 +6,19 @@ use crate::config::{QuorumStoreConfig, SafetyRulesConfig};
 use serde::{Deserialize, Serialize};
 use std::path::PathBuf;
 
+pub(crate) const MAX_SENDING_BLOCK_BYTES_QUORUM_STORE_OVERRIDE: u64 = 4000;
```
Comment (ibalajiarun) [inline config/src/config/consensus_config.rs:9]:
```suggestion
pub(crate) const MAX_SENDING_BLOCK_TXNS_QUORUM_STORE_OVERRIDE: u64 = 4000;
```

File: consensus/src/quorum_store/utils.rs [hunk 2]
```diff
@@ -323,8 +328,23 @@ impl ProofQueue {
...
+        let mut remaining_txns = 0;
+        for (digest, expiration) in self.digest_queue.iter() {
+            // Not expired
+            if *expiration >= current_time {
+                // Not committed
+                if let Some(Some(proof)) = self.digest_proof.get(digest) {
+                    remaining_txns += proof.info().num_txns;
+                }
+            }
+        }
```
Comment (ibalajiarun) [inline consensus/src/quorum_store/utils.rs:341]: This may be fine for now, but we may need a TODO to optimize this, so we don't need to loop through the digest_queue to calculate the number of txns each time.

**Review by ibalajiarun — APPROVED** [review]
LGTM. Left some feedback.

File: config/src/config/quorum_store_config.rs [hunk 3]
```diff
@@ -39,23 +63,21 @@ impl Default for QuorumStoreConfig {
             channel_size: 1000,
             proof_timeout_ms: 10000,
             batch_request_num_peers: 2,
-            mempool_pulling_interval: 100,
-            end_batch_ms: 500,
-            max_batch_counts: 300,
-            max_batch_bytes: 1000000,
+            mempool_pulling_interval: 250,
+            // TODO: This essentially turns fragments off, because there was performance degradation. Needs more investigation.
+            end_batch_ms: 10,
```
Comment (zekun000) [inline config/src/config/quorum_store_config.rs:68]: because back pressure slows down the batch creation?

File: config/src/config/quorum_store_config.rs [hunk 4]
```diff
@@ -39,23 +63,21 @@ impl Default for QuorumStoreConfig {
...
             db_quota: 10000000000,
-            mempool_txn_pull_max_count: 300,
-            mempool_txn_pull_max_bytes: 1000000,
-            // QS will be backpressured if the remaining local batches is more than this number
-            back_pressure_local_batch_num: 10,
+            mempool_txn_pull_max_bytes: 4 * 1024 * 1024,
+            back_pressure: QuorumStoreBackPressureConfig::default(),
             // number of batch coordinators to handle QS Fragment messages, should be >= 1
-            num_workers_for_remote_fragments: 2,
+            num_workers_for_remote_fragments: 10,
```
Comment (zekun000) [inline config/src/config/quorum_store_config.rs:80]: why this change?

File: consensus/src/quorum_store/batch_generator.rs [hunk 5]
```diff
@@ -261,31 +244,52 @@ impl BatchGenerator {
...
                     } else {
-                        counters::QS_BACKPRESSURE.set(0);
-                        // no back pressure
-                        if let Some(proof_rx) = self.handle_scheduled_pull(false).await {
-                            proofs_in_progress.push(Box::pin(proof_rx));
+                        // additive increase, every second
+                        if back_pressure_increase_latest.elapsed() >= back_pressure_increase_duration {
+                            back_pressure_increase_latest = Instant::now();
+                            dynamic_max_pull_count = std::cmp::min(
+                                dynamic_max_pull_count + self.config.back_pressure.dynamic_min_batch_count,
```
Comment (zekun000) [inline consensus/src/quorum_store/batch_generator.rs:283]: this seems not symmetric?

File: consensus/src/quorum_store/utils.rs [hunk 6]
```diff
@@ -35,28 +35,33 @@ pub(crate) struct BatchBuilder {
...
-    pub(crate) fn append_transaction(&mut self, txn: &SignedTransaction) -> bool {
+    pub(crate) fn append_transaction(
+        &mut self,
+        txn: &SignedTransaction,
+        max_txns_override: usize,
+    ) -> bool {
         let serialized_txn = SerializedTransaction::from_signed_txn(txn);
 
-        if self.num_bytes + serialized_txn.len() <= self.max_bytes && self.num_txns < self.max_txns
+        // TODO: bytes, off-by-one?
```
Comment (zekun000) [inline consensus/src/quorum_store/utils.rs:62]: what does this mean?

File: types/src/on_chain_config/consensus_config.rs [hunk 7]
```diff
@@ -70,6 +70,7 @@ impl OnChainConsensusConfig {
 
     pub fn quorum_store_enabled(&self) -> bool {
         match &self {
+            // TODO: this is hardcoded to true, so all continuous forge runs have quorum store enabled
```
Comment (zekun000) [inline types/src/on_chain_config/consensus_config.rs:73]: remove?

**Review by zekun000 — APPROVED** [review]

File: consensus/src/quorum_store/batch_generator.rs (hunk 5 above)
Comment (bchocho) [reply consensus/src/quorum_store/batch_generator.rs:283]: For additive increase, the max_pull_count is increased linearly by the min_batch_count during each duration. This is similar to at least the first versions of TCP.

File: config/src/config/quorum_store_config.rs (hunk 3 above)
Comment (bchocho) [reply config/src/config/quorum_store_config.rs:68]: It shouldn't necessarily, if we're asking for the same number of transactions on the pull. But I was seeing more than one fragment being created. I'll try to revive fragments in the next PR.

File: config/src/config/quorum_store_config.rs (hunk 4 above)
Comment (bchocho) [reply config/src/config/quorum_store_config.rs:80]: This was likely touched when we were doing larger cluster tests, but don't remember how significant this was.

We need to do 400 PoS signatures per second. I'll keep this for now, but we can follow-up with more tuning.
